test/qpy_compat/test_qpy.py

```diff
@@ -311,11 +313,12 @@ def generate_control_flow_circuits():
     circuits.append(qc)
     # While loop
     qc = QuantumCircuit(2, 1)
-
-    with qc.while_loop((qc.clbits[0], 0)):
-        qc.h(0)
-        qc.cx(0, 1)
-        qc.measure(0, 0)
+    block = QuantumCircuit(2, 1)
+    block.h(0)
+    block.cx(0, 1)
+    block.measure(0, 0)
+    while_loop = WhileLoopOp((qc.clbits[0], 0), block)
+    qc.append(while_loop, [0, 1], [0])
```

Comment:
Do you reckon it's worth (in a completely separate issue) adding a global option somewhere to make the builder interface deterministic, purely for testing?  At the time that the circuit is actually constructed, we can do things like sort arguments based on `id` (so it's based on the hash seed) and so forth if we're in deterministic mode.

It'd make some testing easier, but it would add complexity to the builders.

Reply:
Yeah, that's a good idea. I'll open an issue on this now

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/issues/7601